Properly set `has_*` keywords on messages.

aero31aero opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->

This PR takes over from #13196.
**Testing Plan:** <!-- How have you tested? -->
Manual + Automated Tests

**Review Help:**

1. First commit _(has:link)_ by Dustin is mostly unchanged; there are just some refactoring in the tests to make them easier to reuse later.
2. Second commit _(migration)_ by Tim is unchanged.
3. Third commit sets up the way for has:image.
4. Fourth commit _(has:image)_ removes content_has_image as well.
5. Fifth commit _(has:attachment)_ leaves update_calculated_fields as a no-op since it still has a TODO left.
6. Sixth commit removes one of the last occurrences of content_has_attachment.

**Work Remaining:**

1. Remove the last few occurrences of content_has_attachment from the message editing codepaths. This part is slightly tricky as we have to maintain the has_attachment flag from before editing the message and compare it to the final result from bugdown. Naive attempts make the code ugly so I've kept it off the PR for now.
2. Comparing the urls to regexes can still have false positives but a database access to actually check if that url is indeed a valid attachment from Bugdown might not be worth it here. Maybe there's some approach I can't think of.
3. Address the TODO in update_calculated_fields or completely remove the function.

aero31aero commented:
Partially fixes #12912. Tests are passing.

timabbott commented:
Great, thanks for picking this up @aero31aero!  I merged the `self.add_a` commit to help keep the diffs simple.

I think removing `update_calculated_fields` makes sense.  For attachments, I had been thinking that the most correctly implementation would be to have the markdown processor record what URLs there are links to in the message, and then to have the attachment claiming code path inspect those URLs, basically eliminating the regular expression logic from the system (and having `has_attachment` only be true if there's an actual Attachment object associated with the message).

timabbott commented:
Specifically, for the Attachment code path, what I was imagining is changing bugdown to expose a list, `potential_attachment_urls`, on the message object:

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 653bc1889b..eeca4284e4 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -1429,7 +1429,7 @@ def do_send_messages(messages_maybe_none: Sequence[Optional[MutableMapping[str,
 
         # Claim attachments in message
         for message in messages:
-            if message['message'].has_attachment:
+            if len(message['message'].potential_attachment_urls) > 0:
                 do_claim_attachments(message['message'])
 
         for message in messages:
@@ -5452,9 +5452,7 @@ def notify_attachment_update(user_profile: UserProfile, op: str,
     send_event(user_profile.realm, event, [user_profile.id])
 
 def do_claim_attachments(message: Message) -> None:
-    attachment_url_list = attachment_url_re.findall(message.content)
-
-    for url in attachment_url_list:
+    for url in message.potential_attachment_urls:
         path_id = attachment_url_to_path_id(url)
         user_profile = message.sender
         is_message_realm_public = False
@@ -5477,6 +5475,7 @@ def do_claim_attachments(message: Message) -> None:
 
         attachment = claim_attachment(user_profile, path_id, message, is_message_realm_public)
         notify_attachment_update(user_profile, "update", attachment.to_dict())
+        message.has_attachment = True
 
 def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
     old_unclaimed_attachments = get_old_unclaimed_attachments(weeks_ago)
```

With a corresponding update to `check_attachment_reference_change` (potentially a rewrite of that function).  What do you think of that approach?

Also, commit ordering-wise, I think what we want is to move the migration commit to the end of the series.

Finally, it looks like Dustin added `has_link` to `get_raw_db_rows`; is that change correct?  If so, do we need to make similar changes for the other flags?  I think the answer is that change is wrong, and we don't need to add it, because `has_link` and friends aren't accessed by the frontend or intended to be part of the API; they're just stored in the database for use by search.

Otherwise, looks great!

aero31aero commented:
Ready for review. Changes:

- Squashed changes related to has:attachment along with the new potential_urls approach in one commit.
- Removed unneeded changes from the has:link commit.
- Reordered commits and made the migrations commit 'fix' the issue.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -3594,21 +3583,59 @@ def test_claim_attachment(self) -> None:
+        self.send_stream_message(user_profile.email, "Denmark", body, "test")
+        assert_attachment_claimed(dummy_path_ids[2], True)
+
+class MessageHasKeywordsTest(AttachmentTest):
```
Subclassing test classes doens't work how you think it does -- this will run `test_basics` twice.  I don't think we really need a separate class for this bundle of tests.

timabbott commented:
I opened https://github.com/zulip/zulip/issues/13514 for potential follow-up cleanup work (would be a better strategy than the properties, I think, and avoid cluttering Message).

timabbott commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1040,11 +1048,29 @@ def find_proper_insertion_index(self, grandparent: Element, parent: Element,
+            # Populate potential_attachment_urls. do_claim_attachments() handles setting has_attachments.
+            self.markdown.zulip_message.potential_attachment_urls = []
+            for url in unique_urls:
+                self.maybe_append_attachment_url(url)
```
What's your thinking behind filtering out links inside blockquotes here?  I think it's correct for a quoted link to an attachment to still be claimed...

timabbott replied:
Ahh, I see, it's just a bad variable name; I'll make a new `unique_previewable_urls`.

timabbott commented:
This is great; I'm going to just make the changes above and merge this, so we can get it deployed on chat.zulip.org and ready for production use.

timabbott commented:
Merged with a few changes described above (including a test for the blockquote case) and also the addition of 958f39a5516a1671a746ba01c06441ad708e7f4d.  Huge thanks for getting this done in time for the 2.1 release @aero31aero!
